src/module_4/push_exploration_non_linear.md

```diff
@@ -0,0 +1,2005 @@
+
+    # ROC Curve settings
+    plt.subplot(1, 2, 1)
+    plt.plot([0, 1], [0, 1], color="navy", linestyle="--")
+    plt.xlim([0.0, 1.0])
+    plt.ylim([0.0, 1.05])
+    plt.xlabel("False Positive Rate")
+    plt.ylabel("True Positive Rate")
+    plt.title("ROC")
+    plt.legend(loc="lower right", prop={'size': 6})
+    
+    # Precision-Recall Curve settings
+    plt.subplot(1, 2, 2)
+    plt.xlim([0.0, 1.0])
+    plt.ylim([0.0, 1.05])
+    plt.xlabel("Recall")
+    plt.ylabel("Precision")
+    plt.title("Precision-Recall Curve")
+    plt.legend(loc="upper right", prop={'size':6})
+
+    plt.tight_layout()
+
+    if save_curves_path:
+        # Define the filename with a timestamp
+        figure_filename = f"Evaluation_Curves_Multiple_Models_{timestamp}.png"
+        figure_path = os.path.join(save_curves_path, figure_filename)
+
+        plt.savefig(figure_path)
+
+    plt.show()
+```
+
+
+```python
+# Train regression tree
+tree = DecisionTreeClassifier()
+tree.fit(X_train, y_train)
+y_pred_tree = tree.predict(X_val)
+
+# Train logistic regression
+lr = LogisticRegression(penalty="l2", C=1e-5, solver="liblinear")
+lr.fit(X_train, y_train)
+y_pred_lr = lr.predict_proba(X_val)[:, 1]
+
+generate_evaluation_curves_multi(
+    model_names=["ClassifierTree", "LogisticRegression"],
+    y_preds=[y_pred_tree, y_pred_lr],
+    y_test= y_val
+)
+
+```
+
+
+    
+![png](push_exploration_non_linear_files/push_exploration_non_linear_14_0.png)
+    
+
+
+Obviously, a single decision tree will hardly beat it. So we will try using random forests with different params of n_trees. As as curiosity, if we take a look at the plot for LR with same parameters as trained before, the AUC for PR is 0.16 vs 0.17 before (C, solver and  regularisation remains equal). I think this might happen as model regularisation can still have some randomness in model weights, hence getting different results. For addressing that we could have made use of random_seed.
+
+## Random Forests
+
+
+```python
+# Train Logistic Regression
+lr = LogisticRegression(penalty="l2", C=1e-5, solver="liblinear")
+lr.fit(X_train, y_train)
+y_pred_lr = lr.predict_proba(X_val)[:, 1]
+
+# Train RandomForestClassifier for various n_trees
+trained_models = []
+y_preds_rf = []
+
+n_trees_values = [5, 20, 50, 100, 250]
+for n_trees in n_trees_values:
+    rf = RandomForestClassifier(n_estimators=n_trees)
+    rf.fit(X_train, y_train)
+    y_preds_rf.append(rf.predict_proba(X_val)[:, 1])
+```
+
+
+```python
+model_names = ["LogisticRegression"] + [f"RandomForest (n_trees={n})" for n in n_trees_values]
+y_preds = [y_pred_lr] + y_preds_rf
+
+generate_evaluation_curves_multi(model_names, y_preds, y_val)
+```
+
+
+    
+![png](push_exploration_non_linear_files/push_exploration_non_linear_18_0.png)
+    
+
+
+For ROC we see no improvement, whereas higher number of trees works better than just a few. This model didn't improve our baseline defined by the linear model in blue.
```

Comment:
This is defo a little bit weird (that RF are not superior to LR), maybe it'd be interesting to play a little bit more with it. specially tune the parameters of the trees and look what happens in train vs test